doc/nnet.1
.\" Process this file with
.\" groff -man -Tascii nnet.1
.\"
.\" Copyright (c) 2016, Ray Dillinger <bear@sonic.net>
.\" Copyright (c) 2022, Karl Semich <0xloem@gmail.com>
.\"
.\" %%%LICENSE_START(AGPLv3+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU Affero General Public License as
.\" published by the Free Software Foundation; either version 3 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU Affero General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Affero General Public License for more details.
.\"
.\" You should have received a copy of the GNU Affero General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END
.\"
.TH NNET 1 "JANUARY 2017" Linux "User Manuals"
.SH NAME
nnet \- define, train, and deploy feedforward, recurrent, or deep neural networks.
.SH SYNOPSIS
.B nnet
.I filename

.B nnet -v, nnet --version

.B nnet -h, nnet --help

.B nnet -l, nnet --language

.SH DESCRIPTION
.B nnet
is a utility for general development of neural networks.  The neural networks it defines are not limited as to topology
or depth, and many unusual node types such as LSTMs, pooling nodes, etc, are supported.  When invoked with a
.I filename
argument,
.B nnet
processes
.I filename.nnet
which must be a nnet script. See
.B nnet_language(1)
or invoke
.B nnet -l
for information about the script language. Nnet scripts specify the topology, weights, firing sequence, and data of an
artificial neural network.  Optionally they may also specify training and/or testing instructions and data (or
locations of data sources and destinations) for training, testing, validation, or deployment.

When finished processing, or at intervals during training as specified by any training instructions in the script,
.B nnet
writes
.I filename.out,
a nnet script that specifies the network including changes made during processing or training.  If
.I filename.out
is saved while training is not yet complete, it will include a reduced training schedule corresponding to the amount of
training that remains to be done.  If training is interrupted, renaming
.I filename.out
to
.I filename.nnet
and restarting with
.B nnet
.I filename
will resume training from the most recent save point.

Nnet scripts may be edited in any text editor.  Users may use the script language to add new training, testing, or
deployment instructions, to change training parameters or switch to a different training method, or to directly modify
the network definition and topology, eg by changing the number of input, output, or hidden nodes, adding layers while
building deep networks, adding new connections or removing existing connections, etc. The edited file may then be saved
as
.I filename.nnet
and used to restart
.B nnet
for continued processing.  Topology or weight edits will usually make the network less accurate until retraining has
taken place, but whatever training has already been done need not be entirely lost.

.SH OPTIONS
.IP -v,--version
Print the current version information of nnet and exit.
.IP -h,--help
Print this man page and exit.
.IP -l,--language
Print the nnet_language man page and exit.

.SH
.SH REPORTING BUGS
.B nnet
is part of the gneural_network project. You may report any bugs, or review the bug database, at

<https://savannah.gnu.org/bugs/?group=gneuralnetwork>.

.B nnet
contains numerous self-diagnostic and constraint checks for things which will never happen unless the code contains an
unsuspected bug.  Please notify us immediately if it ever exits printing a diagnostic message starting with the
words 'Program Error.' Any such message would be reporting a bug in
.B nnet
itself, not a problem with your script.
.SH AUTHOR
Man page written by Ray Dillinger <bear at sonic dot net>
.SH COPYRIGHT
.B nnet
and its documentation are Copyright (C) 2016-2022 Karl Semich, Ray Dillinger, Jean-Michel Sellier, and the Gneural_Knockoff project.

Permission is granted to modify and distribute the software and its sources subject to the conditions of the GNU Public
License, version 3 or later.  Permission is granted to distribute the documentation, including this man page, subject to
the conditions of the GNU Free Documentation License, version 1.3 or later.

.SH SEE ALSO
.BR nnet_language (1)


NEWS
--- gneural_knockoff release 0.9.2

    [day] Sep 2022, [name], 0xloem@gmail.com

    A new release is not yet ready.

    I have put some work into stabilising the existing code more. I would
    prefer to merge the two parsing and processing systems together.

    I hope more contributions may happen.

--- gneural_network release 0.9.1

    03 May 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- The parser has been drastically improved.

    --- The code has been restructured to avoid the use of global variables,
        making this package easier to understand and maintain.

--- gneural_network release 0.9.0

    13 Apr. 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- A bug has been fixed in the computation of the neural
        network error.

    --- The multi-scale Monte Carlo algorithm to train a
        neural network has been parallelized by means of
        OpenMP.

    --- The genetic algorithm to train a neural network has
        been parallelized by means of OpenMP.

    --- The script modulus_of_two_dimensional_vectors.input
        has been fixed.


--- gneural_network release 0.8.0

    02 Apr. 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- The previous (simplified) genetic algorithm has been
        drastically improved.
    --- A multi-scale Monte Carlo optimizer has been implemeted
        to train the neural network.
    --- New scripts have been implemented and are provided in
        the folder "tests", one which fits a quandratic curve,
        another which trains a network to compute the modulus
        of a two-dimensional vector and, finally, one which
        trains a network to invert the order of numbers in a
        given couple (a,b).
    --- Several bugs have been fixed.

--- gneural_network release 0.6.0

    24 Mar. 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- The code has been divided into proper header and
        implementation files.
    --- The code and the scripting language have been
        generalized to handle a general number of neurons in
        both the input and output layers.
    --- A neural network can now be saved in a file specified by
        the user (to be loaded afterwards).
    --- Some part of the code have been further optimized
        to run faster.
    --- The user can now decide if the initial weights of a
        network have to be randomized before the training.
    --- Several important bugs have been fixed.

--- gneural_network release 0.5.0

    18 Mar. 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- The installation process has been standardized as
        it should be for a GNU package.
    --- A scripting language has been implemented which allows
        the user to define his/her own neural network without
        having to go through the source code.
        Plenty of options are now possible, for instance, the
        user can define the number of neurons, their connections,
        their accumulator, activation functions, etc.
        The user can choose among four different methods for
        training the neural network which can be saved for
        later use.
    --- Some part of the code have been optimized to run faster.
    --- A test folder has been created containing a commented
        example of script which defines a neural network for
        a simple curve fitting task.

--- gneural_network release 0.0.1

    03 Mar. 2016, J.M. Sellier, jeanmichel.sellier@gmail.com

    --- First release of gneural_network



src/nnet.c
/*
   Copyright (C) 2022 Karl Semich <0xloem@gmail.com>
   Copyright (C) 2016-2017 Ray Dillinger <bear@sonic.net>

   This file is part of Gneural Knockoff.

   Gneural Knockoff is free software; you can redistribute it and/or modify it
   under the terms of the GNU Affero General Public License as published by the
   Free Software Foundation; either version 3, or (at your option) any later
   version.

   Gneural Knockoff is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
   License for more details.

   You should have received a copy of the GNU Affero General Public License
   along with Gneural Knockoff.  If not, see <http://www.gnu.org/licenses/>.
*/




#include "includes.h"
#include "defines.h"
#include "parser.h"
#include "save.h"

#define HELPSTRING  "usage: nnet <filename> | nnet -v | nnet -h | nnet -H | nnet -l \nOptions:\n\
  -h, -?, --help:  print this help and exit.\n\
  -H, --manpage:   print a manual fully describing nnet.\n\
  -l, --language:  print a manual describing the nnet script language.\n\
  -v, --version:   version and copyright information.\n"

int
HandleOptions (int argc, char **argv)
{                               // name, args, NULL, returnval
  static struct option options[] = {
    {"help", no_argument, NULL, 'h'}, {"version", no_argument, NULL, 'v'},
    {"manpage", no_argument, NULL, 'H'}, {"language", no_argument, NULL, 'l'},
    {"?", no_argument, NULL, '?'}, {0, 0, 0, 0}
  };
  int opt;
  opterr = 0;
  while ((opt = getopt_long (argc, argv, "hvHl", options, NULL)) != -1)
    switch (opt)
      {
      case 'v':
        printf
          ("nnet 0.0.1\nCopyright(C) 2016-2017 gneural_network developers\nLicense LGPLv3+. For information about copying, modifying, "
           "and distribution see <http://gnu.org/licenses/lgpl.html>.\n");
        return (0);
      case 'H':
        printf (MANPAGE);
        return (0);
      case 'l':
        printf (LANGMAN);
        return (0);
      default:
        printf (HELPSTRING);
        return (0);
      }
  return (1);
}

// objectives: establish name for reading and default nonserialized name for writing.  readname should be arg (plus .nnet extension if .nnet extension isn't
// already present) config->savename should be arg (without serial number and without .nnet extension), plus .out extension.  serial number if present in
// filename should be saved in config, else zero should be saved in config.
void
GetFileNames (char *readname, struct conf *config, const char *arg)
{
  config->flags |= SAVE_DEFAULT;
  size_t index = 0;
  size_t serialstart = 0;
  size_t serialend = 0;
  size_t serial = 0;
  size_t len = strlen (arg);
  for (index = 0; serialend == 0 && index < len && index < 255; index++)
    {
      if (arg[index] == '.' && serialstart == 0)
        serialstart = index + 1;
      else if (isdigit (arg[index]) && serialstart != 0 && serialend == 0)
        {
          serial *= 10;
          serial += arg[index] - '0';
          if (arg[index + 1] == '.' || arg[index + 1] == 0)
            serialend = index;
        }
      else if (serialstart != 0 && serialend == 0)
        {
          serialstart = 0;
          serial = 0;
        }
    }
  config->serialnum = serial;
  config->savename = malloc (sizeof (char) * (len + 25));
  if (config->savename == NULL)
    {
      fprintf (stderr, "allocation failure atempting to save filename.\n");
      exit (1);
    }
  strcpy (readname, arg);
  if (strcmp (".nnet", &(arg[len - 5])) == 0)
    readname[len - 5] = 0;
  strcat (readname, ".nnet");
  len = strlen (readname) - 5;
  if (serialstart != 0)
    {
      for (index = 0; index < serialstart; index++)
        config->savename[index] = readname[index];
      for (index = 1; serialend + index < len; index++)
        config->savename[serialend + index] = readname[index];
      config->savename[serialend + index + 1] = 0;
    }
  else
    {
      strcpy (config->savename, readname);
      config->savename[len] = 0;
    }
  strcat (config->savename, ".out");
}


// if not serializing, filename = config->savename.
// if serializing & there is a dot in the filename, last dot in filename is replaced with dot,++serial,dot
// if serializing & there is no dot in the filename, the filename is extended with dot-serial.
// if serializing, increment serial number in config.
// filename is a 256-char buffer.
void
NameOutputFile (char *filename, struct conf *config)
{
  size_t lastdot = 0;
  size_t len = strlen (config->savename);
  size_t index = len;
  size_t writeindex = 0;
  if ((config->flags & SAVE_SERIALIZE) == 0)
    {
      strncpy (filename, config->savename, 255);
      return;
    }
  for (index = 1; index <= len && config->savename[len - index] != '.';
       index++);
  lastdot = (config->savename[len - index] == '.') ? len - index : len;
  for (index = 0; (index <= len) && writeindex < 256; index++)
    if (index != lastdot)
      {
        filename[writeindex++] = config->savename[index];
        if (index + 1 == len && lastdot == len)
          writeindex +=
            snprintf (&(filename[writeindex]), 255 - index, ".%d",
                      ++(config->serialnum));
      }
    else
      writeindex +=
        snprintf (&(filename[writeindex]), 255 - index, ".%d.",
                  ++(config->serialnum));
  filename[writeindex] = 0;
}

int
main (int argc, char **argv)
{
  struct nnet newt;
  struct conf netconf;
  struct slidingbuffer bf;
  char fname[256];
  fname[0] = 0;
  char *filename = &(fname[0]);
  if (!HandleOptions (argc, argv))
    exit (0);
  if (argc > 2)
    fprintf (stderr,
             "%s does not process more than one nnet script in a single invocation.\n",
             argv[0]);
  if (argc != 2)
    {
      fprintf (stderr,
               "Usage:  %s [filename] where 'filename.nnet' is the name of a nnet script.\n",
               argv[0]);
      exit (1);
    }
  memset (&newt, 0, sizeof (struct nnet));
  memset (&bf, 0, sizeof (struct slidingbuffer));
  memset (&netconf, 0, sizeof (struct conf));
  GetFileNames (filename, &netconf, argv[1]);
  if (NULL == (bf.input = fopen (filename, "r")))
    {
      fprintf (stderr, "unable to open %s\n", filename);
      exit (1);
    }
  nnetparser (&newt, &netconf, &bf);
  fflush (stdout);
  PrintWarnings (&bf);
  if ((netconf.flags & SILENCE_DEBUG) != 0)
    {
      printf ("\n\nParse successful.");
      printf ("\n%d nodes created. Node 0 is a bias node.  ", newt.nodecount);
      if (newt.inputcount > 0)
        printf ("Nodes {1 %d} are input nodes.  ", newt.inputcount - 1);
      if (newt.nodecount - newt.inputcount - newt.outputcount > 0)
        printf (" Nodes {%d %d} are hidden nodes.", newt.inputcount,
                (newt.nodecount - 1) - newt.outputcount);
      if (newt.outputcount > 0)
        printf (" Nodes {%d %d} are output nodes. ",
                newt.nodecount - newt.outputcount, (newt.nodecount - 1));
      printf ("\n%d Connections created. ", newt.synapsecount);
      for (int syn = 0; syn < newt.synapsecount; syn++)
        printf ("%d=%d->%d,", syn, newt.sources[syn], newt.dests[syn]);
      printf ("\b \n");
    }
  fclose (bf.input);
  bf.input = NULL;
  NameOutputFile (filename, &netconf);
  FILE *outf = fopen (filename, "w");
  if (outf == NULL)
    {
      fprintf (stderr, "unable to open %s", filename);
      exit (1);
    }
  if ((netconf.flags & SILENCE_DEBUG) != 0)
    debugnnet (&newt);
  nnetwriter (&newt, &netconf, outf);
  fclose (outf);
}
